### ISSUE TITLE
Vulkan Forces >60hz Monitor to Switch to 60hz (Windows-Nvidia)

### ISSUE THREAD

**Issue by Awakened0** [libretro/RetroArch#3494]
Trying to nail down reproducing this one is maddening. I'm using a 120hz monitor and I like running it at native 120hz with swap interval 2. The hz change seems to happen mostly from running games in the latest MAME core or in an older 2015 core I usually use. But I've had the problem "leak" into a game running in Genesis Plus GX after testing several MAME games and having to reset my hz back to 120 in my Windows display control panel between MAME game launches. Before the problem "leaked" from MAME I was able to launch into a game in GPGX just fine. Then, when both the MAME and GPGX games I was testing were both forcing 60hz, I launched a game in Mednafen Saturn, which didn't force 60hz. Then I launched both the GPGX game and the MAME game and then neither forced 60hz! But then I launched a different MAME game and it forced it again...arg.

I usually use windowed fullscreen. Switching to true fullscreen didn't stop the hz change, but at least in that mode when I exited RetroArch it switched back to 120 automatically.

I can't get a log with this problem since it doesn't ever happen when I use retroarch_debug.exe with log verbosity enabled. It happened once when I had verbosity disabled and with the same frequency as retroarch.exe with the -D switch, but otherwise it won't force the hz change. Maybe something about the command window delaying the RA window coming up for an extra second while it prints the verbose text?

I tried setting the preferred refresh rate to Maximum in my NVidia control panel instead of Application Controled, but that didn't mitigate the issue.

I suppose this could be another teething issue with the Vulkan drivers, but I figured I'd post it here just in case a dev might have an idea on how to fix it on the RA end. I think I've had this happen once in GL with Reicast, but haven't had it happen since in that driver.

**Comment by Awakened0** [comment]
I just discovered that if I set swap interval to 1 and turn off windowed full screen, when I launch a game it sets the refresh to 60hz every time. It's only flaky about it with swap interval 2 with either windowed or true full screen. It's surprising it can change the refresh rate in windowed full screen. Maybe something with how swap interval is initialized is causing the refresh rate change?

**Comment by Awakened0** [comment]
This now happens consistently across all cores. And different swap interval settings doesn't change the behavior. I'm pretty sure as NVidia drivers updated it became consistent. I'm on 378.66 currently.

The difference between windowed and true fullscreen is that when this happens in windowed, the screen stays black when you quit RA. I found out if I use Win+P to switch to my other display and back, I can get out of the black screen. In true fullscreen mode, when you quit it switches back to 120hz properly with no black screen.

At least one other person (I think Saftle?) confirmed the issue on IRC, as well as JackOfOwls in this forum post: https://forums.libretro.com/t/cant-exit-mednafen-psx-hc-full-screen-without-hard-crash/8578

**Comment by Silanda** [comment]
FWIW, I'm think getting this issue too. Windows 10, GTX 1080 (used to happen on my 780ti too). If I use the Vulkan renderer in windowed fullscreen mode at any refresh other than 60.0 Hz, the refresh tends to change to 60 and I'm met with a black screen on alt-tab or exit. I have to log out blind to get the display back. Hitting fast forward in Mednafen PSX HW always triggers this. I'm not running at 120 Hz, I either run 59.94 or 50 Hz. Anything other than exactly 60 triggers it. Exclusive fullscreen isn't really a solution for me as it too switches to 60Hz (using Vulkan or OpenGL), which isn't very good for PAL games.

I'm guessing Nvidia driver bug that Retroarch in particular triggers. I used to have a similar problem in OpenGL mode for a while too (refresh change, black screen on exit), but that no longer happens.

**Comment by twinaphex** [comment]
@Themaister mentioned something about being able to use VK_KHR_display as a context driver with Nvidia. I am not sure if this can work for Windows too. Basically, the hope is that using this context driver could give you the best possible experience.

**Comment by twinaphex** [comment]
I think these issues you are fighting are entirely driver bound and there is not really much we can do.

**Comment by Silanda** [comment]
Just thought I'd post a quick follow up. While I can't be absolutely certain that something else hasn't caused this, since the Windows Creators Update I'm no longer getting the loss of picture described earlier, so perhaps that was a Windows bug. The refresh rate still changes as described however.

**Comment by Awakened0** [comment]
I was testing Vulkan today and found that with the 381.65 drivers if swap interval is set to 2 it doesn't force my 120hz monitor into 60hz mode in either borderless or true fullscreen modes. At least on launch. If I press the fast forward key, then it forces it into 60hz. Swap interval 1 forces it to 60hz at launch, which is the same behavior as GL in true fullscreen.

**Comment by Awakened0** [comment]
With today's nightly, Vulkan forces 60hz on the 388.31 driver I'm using as soon as I start any content in RetroArch. It happens when using either swap interval 1 or 2 in either true or windowed fullscreen modes. It's obviously not a bad thing when using swap interval 1, since that will give smooth scrolling at 60hz. If I use true fullscreen it will even switch back to 120hz like it should after exiting RA. But, I prefer to use 120hz with swap interval 2, since my monitor doesn't have to switch modes and it looks brighter and I think has less lag at it's native refresh. The OGL driver never forces 60hz.

**Comment by Awakened0** [comment]
With today's nightly and the 390.77 driver it now properly leaves my monitor running at 120hz when swap interval 2 is set.

**Comment by Fergdog** [comment]
The Vulkan issue of changing a monitor from 120hz to 60hz while fullscreen came back with 1.7.4. Sorry for not testing it earlier. 1.7.3 has no issues still. Between 1.7.3 and 1.7.4 something broke 120hz Vulkan on Nvidia. 1.7.5 also has the same issues.

As soon as I open a beetlepsx game with vulkan, the framerate gets really choppy. I have vsync interval set to 2 to half my monitors refresh rate but because it switches my monitor to 60hz it behaves as if it were a 30hz monitor. Changing the swap interval is nearly instantaneous on 1.7.3 and doesn't change my monitors refresh rate.

I posted this twice in other threads but should have posted it here first sorry. I'll delete the others.

**Comment by Awakened0** [comment]
Yeah, my new monitor just goes into GSync mode and adjusts to the content's framerate with Vulkan in true fullscreen and the "Sync to Exact Content Framerate" throttle option on. So I don't need to use swap interval 2 anymore.

**Comment by Fergdog** [comment]
Great, thanks for leaving me behind jerk! I'm still stuck with a non gsync 120hz monitor hah

**Comment by Fergdog** [comment]
That was merged on September 8th while 1.7.4 came out on August 30th so it must be something sooner, unless parts of that were merged sooner.

**Comment by Fergdog** [comment]
@Themaister Could you take a look at this please?

**Comment by Themaister** [comment]
I don't have a 120Hz monitor. I can't fix issues I can't repro.

**Comment by Awakened0** [comment]
@Fergdog Try the latest nightly. There is a Screen Resolution option under Settings, Video where you can select different resolutions and refresh rates. If you use that to manually pick your screen's native res and refresh, it should keep your refresh at that.

**Comment by Fergdog** [comment]
Nah it doesn't fix it, you don't even have to bootup a game to test it. All you have to do is change the the vsync swap interval and screen flickers for a bit and changes to 60hz and sometimes changes it back to 120hz after closing RA and sometimes doesn't.

**Comment by Ryusennin** [comment]
For info, I've been researching this issue for months. I don't think it can be fixed.

On Windows 7, you can force Vulkan to run in _any_ refresh rate (I tried 50, 100, 120, 144).

On Windows 10, Vulkan is always stuck to 60 Hz in all apps/emus/games I could test (RetroArch, RPCS3, Detroit, RDR2, dx11 games running via dxvk). It happens both on nVidia and AMD cards, so it's not specific to one driver. Forcing the app in 50 Hz for example will result in a blank screen (although the game clearly runs in the background).
